aeneas/src/ir/SsaNormalizer.v3

```diff
@@ -346,6 +346,306 @@ class SsaRaNormalizer extends SsaRebuilder {
 				var nobj = if(nargs.length == 1, newGraph.nullConst(ptrType), nargs[1]);
 				return mapN(app, [nfunc, nobj]);
 			}
+			RefStructAt => {
+				var an = ArrayNorm.!(normType(V3.arrayByteType));
+				var newArgs = genRefs(app.inputs);
+				if (!app.facts.O_NO_BOUNDS_CHECK) {
+					// XXX: manually CSE the promotions here?
+					var array = newArgs[0], offset = newArgs[an.size];
+					curBlock.opBoundsCheck(an.newType, array, offset);
+					var t = RefType.!(op.typeArgs[0]);
+					var struct_size = newGraph.intConst(t.structDecl.byteSize);
+					var end_index = curBlock.opIntAdd(offset, struct_size);
+					var array_length = curBlock.opArrayGetLength(an.newType, array);
+					var c1 = curBlock.opIntULt(Int.TYPE, Int.TYPE, array_length, end_index);
+					curBlock.opConditionalThrow(V3Exception.BoundsCheck, c1);
+				}
+				return mapN(app, newArgs);
+			}
+			RefStructIn(offset) => {
+				var an = ArrayNorm.!(normType(V3.arrayByteType));
+				var newArgs = genRefs(app.inputs);
+				var array = newArgs[0], pos = newArgs[an.size];
+				newArgs[an.size] = curBlock.opIntAdd(newGraph.intConst(offset), pos);
+				return mapN(app, newArgs);
+			}
+			RefStructGetField(offset) => {
+				var an = ArrayNorm.!(normType(V3.arrayByteType));
+				var ortype = normReturnType(op).oldType, nrtype = normReturnType(op).newType;
+				var newArgs = genRefs(app.inputs);
+				var array = newArgs[0], pos = newArgs[an.size];
+				var result = SsaInstr.!(newGraph.intConst(0));
+				match (ortype) {
+					x: IntType => {
+						var width = IntType.!(ortype).packedByteSize();
+						// assume little endianess
+						for (i < width) {
+							var read = curBlock.opArrayGetElem(an.newType, Int.TYPE, app.facts,
+								newArgs[0], curBlock.opIntAdd(newGraph.intConst(i + offset), pos));
+							var shifted = curBlock.addApply(app.source, IntType.!(ortype).opShl(),
+								[read, newGraph.intConst(8 * i)]);
+							result = curBlock.addApply(app.source, IntType.!(ortype).opOr(),
+								[shifted, result]);
+						}
+						return map1(app, result);
+					}
+					x: EnumType => {
+						var width = EnumType.!(ortype).packedByteSize();
+						for (i < width) {
+							var read = curBlock.opArrayGetElem(an.newType, Int.TYPE, app.facts,
+								newArgs[0], curBlock.opIntAdd(newGraph.intConst(i + offset), pos));
+							var longViewOp = V3Op.newIntViewI(read.getType(), Long.TYPE);
```

Comment:
Instead of Long.TYPE, I think you want to use the enum's tag type, either from `V3.getVariantTagType` or from `x.enumDecl.tagType`.

Reply:
I still used Int.TYPE since there might be the case that the value read in is greater than the maximum of tag value and we need to return an oob 0. The type of the tag is usually too short. Suppose we're reading in 0x128 for a enum type with maximum tag value 2, we should be returning 0, but if we use the tag's type, we'll fail to catch this oob.